File: app/api/spelling.py
"""
app/api/spelling.py — Spelling & Literacy API

Serves the literacy orchestrator payload.
"""

from fastapi import APIRouter
from app.services.spelling.literacy_orchestrator import generate_literacy_session_config
from app.services.spelling.spelling_state_machine import SpellingProgression
from pydantic import BaseModel

router = APIRouter()

@router.get("/config")
def get_spelling_config(word: str, learning_style: str = "adult", profile_type: str = "child"):
    """Returns the full spelling orchestrator payload for the frontend."""
    return generate_literacy_session_config(word, learning_style, profile_type)

class SpellingAttemptRequest(BaseModel):
    target_word: str
    current_progress: list[str]
    attempted_letter: str

@router.post("/attempt")
def validate_spelling_attempt(req: SpellingAttemptRequest):
    """
    Stateless spelling progression validation.
    Returns the new spelling state and next expected letter.
    """
    prog = SpellingProgression(req.target_word)
    prog.progress = req.current_progress
    
    return prog.process_attempt(req.attempted_letter)

File: frontend/src/api/spelling.ts
/**
 * spelling.ts — API Client for Phase 6.3 Spelling System
 */

import { api } from "./client";

export interface WordMetadata {
  word: string;
  difficulty: number;
  phonics?: string;
  stroke_complexity?: string;
  focus_letters: string[];
  category?: string;
}

export interface LiteracySessionConfig {
  word_metadata: WordMetadata;
  assist_profile: {
    show_ghost_word: boolean;
    show_next_letter_hint: boolean;
    auto_repeat_pronunciation: boolean;
    forgiveness_level: string;
  };
  coaching_policy: any;
  audio: {
    pronunciations: {
      word: string;
      letters: Record<string, string>;
    };
    coach_audio: Record<string, string>;
  };
}

export interface SpellingProgressionState {
  target: string;
  progress: string[];
  next_expected: string | null;
  state: "idle" | "listening" | "attempting" | "partial_success" | "correct" | "retrying" | "celebrating";
  accuracy: number;
  attempts_on_current: number;
}

export const spellingApi = {
  /** Fetch the complete literacy orchestrator payload for a specific word. */
  getConfig: (word: string, learningStyle: string = "adult", profileType: string = "child") => {
    return api.get<LiteracySessionConfig>(`/spelling/config`, {
      params: { word, learning_style: learningStyle, profile_type: profileType }
    });
  },

  /** Validates a letter attempt and returns the updated state machine. */
  validateAttempt: (targetWord: string, currentProgress: string[], attemptedLetter: string) => {
    return api.post<SpellingProgressionState>("/spelling/attempt", {
      target_word: targetWord,
      current_progress: currentProgress,
      attempted_letter: attemptedLetter
    });
  }
};
